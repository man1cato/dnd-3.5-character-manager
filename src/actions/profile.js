import database from '../firebase/firebase'


//SET PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
});

//READ PROFILE FROM FIREBASE
export const startSetProfile = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val();
            dispatch(setProfile(id, profile));
        });
    };
};

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
});

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates));
        });
    };
};

//DELETE PROFILE FROM STORE
export const deleteProfile = () => ({
    type: 'DELETE_PROFILE'
})

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
