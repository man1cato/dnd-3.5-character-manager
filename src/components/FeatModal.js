import React from 'react'
import Modal from 'react-modal'

if (process.env.NODE_ENV !== 'test') { Modal.setAppElement('#app') }

const FeatModal = ({selected, handleCloseModal}) => (
    <Modal
        isOpen={!!selected}
        onRequestClose={handleCloseModal}
        contentLabel="Selected Feat"
        className="modal"
    >
        {selected && 
            <div>
                <h4>{selected.name}</h4>
                <h5>{selected.types.join(", ")}</h5>
                <p>{selected.description}</p>
            </div>
        }
    </Modal>

)

export default FeatModal