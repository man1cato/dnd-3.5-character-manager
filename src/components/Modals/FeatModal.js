import React from 'react'

import DefaultModal from './DefaultModal'


const FeatModal = ({ clickedFeat, handleCloseModal }) => (
    <DefaultModal
        clickedItem={clickedFeat}
        onRequestClose={handleCloseModal}
        contentLabel="Selected Feat"
    >
        {clickedFeat && 
            <>
                <h3>{clickedFeat.name}</h3>
                <h5>{clickedFeat.types.join(", ")}</h5>
                <p>{clickedFeat.description}</p>
            </>
        }
    </DefaultModal>
)

export default FeatModal