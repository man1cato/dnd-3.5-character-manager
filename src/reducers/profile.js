export default (state = {}, action) => {
    switch (action.type) {
        case 'SET_PROFILE':
            return {
                id: action.id,
                ...action.fields
            };
        case 'EDIT_PROFILE':
            return {
                ...state,
                ...action.updates
            };
        default:
            return state;
    }
};
