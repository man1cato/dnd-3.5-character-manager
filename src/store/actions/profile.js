import database from '../../firebase/firebase'
import { startGetProfiles } from './profiles'

//SET SELECTED PROFILE IN STORE
export const setProfile = (id, profile) => ({
    type: 'SET_PROFILE',
    id,
    profile
})

//READ PROFILE FROM FIREBASE
export const startSetProfile = (uid, id) => {
    return (dispatch) => {
        return database.ref(`users/${uid}/profiles/${id}`).once('value').then((snapshot) => {
            const profile = snapshot.val()
            dispatch(setProfile(id, profile))
        })
    }
}

//CREATE PROFILE IN FIREBASE
export const startCreateProfile = (profile) => {    
    return (dispatch, getState) => {
        const uid = getState().auth.uid       
        return database.ref(`users/${uid}/profiles`).push(profile).then((ref) => {
            dispatch(setProfile(ref.key, profile))
            dispatch(startGetProfiles(uid))
            localStorage.setItem('selectedCharacterId', ref.key)
        })
    }
}


//UPDATE PROFILE IN STORE
export const editProfile = (updates) => ({
    type: 'EDIT_PROFILE',
    updates
})

//UPDATE PROFILE IN FIREBASE
export const startEditProfile = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid
        return database.ref(`users/${uid}/profiles/${id}`).update(updates).then(() => {
            dispatch(editProfile(updates))
        })
    }
}

//REMOVE PROFILE FROM STORE
export const removeProfile = () => ({
    type: 'REMOVE_PROFILE'
})

//REMOVE PROFILE FROM FIREBASE
// export const startRemoveProfile = () => {
//     return (dispatch) => {
//         return database.ref(`users/${uid}/profiles/${id}`).remove().then(() => {
//             dispatch(removeProfile())
//         })
//     }
// }

//LEVEL UP CHARACTER IN STORE
// export const levelUp = (character) => ({
//     type: 'LEVEL_UP',
//     character
// });
