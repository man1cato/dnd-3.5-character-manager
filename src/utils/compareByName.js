export default (a,b) => {
    const aName = a.name.toUpperCase();
    const bName = b.name.toUpperCase();
    if (aName < bName) { return -1 };
    if (aName > bName) { return 1 };
    return 0;
}