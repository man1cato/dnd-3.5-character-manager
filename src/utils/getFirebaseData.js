import admin from 'firebase-admin'
import serviceAccount from '../../config/dnd-3-5-character-manager-e3f58e584d80.json'

admin.initializeApp({
   credential: admin.credential.cert(serviceAccount),
   databaseURL: "https://dnd-3-5-character-manager.firebaseio.com"
 })
 
const database = admin.database()

export default async () => {
   try {
      const snapshot = await database.ref('api').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getRaces = async () => {
   try {
      const snapshot = await database.ref('api/races').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getJobClasses = async () => {
   try {
      const snapshot = await database.ref('api/jobClasses').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getSkills = async () => {
   try {
      const snapshot = await database.ref('api/skills').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getFeats = async () => {
   try {
      const snapshot = await database.ref('api/feats').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getSpecialAbilities = async () => {
   try {
      const snapshot = await database.ref('api/specialAbilities').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getItems = async () => {
   try {
      const snapshot = await database.ref('api/items').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}

export const getSpells = async () => {
   try {
      const snapshot = await database.ref('api/spells').once('value')
      return snapshot.val()
   } catch (e) {
      console.error(e)
   }
}