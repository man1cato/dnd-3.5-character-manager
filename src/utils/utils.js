import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])