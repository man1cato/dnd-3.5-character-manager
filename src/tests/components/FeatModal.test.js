import React from 'react'
import {shallow} from 'enzyme'

import FeatModal from '../../components/FeatModal'
import profile from '../fixtures/profile'
import { getFeats } from '../../utils/getFirebaseData'

const handleCloseModal = jest.fn()

const props = {
   selected: {
      name: "Manyshot",
      description: "Shoot two or more arrows simultaneously",
      types: ["General", "Fighter Bonus"]
   },
   handleCloseModal
}

test('should render FeatModal', () => {
   const wrapper = shallow(<FeatModal {...props} />)
   expect(wrapper).toMatchSnapshot()
})