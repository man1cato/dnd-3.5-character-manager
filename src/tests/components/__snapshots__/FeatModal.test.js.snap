// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render FeatModal 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={0}
  contentLabel="Selected Feat"
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  role="dialog"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div>
    <h4>
      Manyshot
    </h4>
    <h5>
      General, Fighter Bonus
    </h5>
    <p>
      Shoot two or more arrows simultaneously
    </p>
  </div>
</Modal>
`;
